### ISSUE TITLE
Cannot deploy the document site

### ISSUE THREAD

**Issue by Atry** [hhvm/user-documentation#1222]
According to [the Github Action log](https://github.com/hhvm/user-documentation/runs/6699209905?check_suite_focus=true), we encountered the following error when deploying a new version of this document site:

```
Instance deployment: The Docker container unexpectedly ended after it was started.
```

I investigated the events on the AWS web console, and it seems what get killed is not only the newly deployed site, but also the previously deployed server (in an existing Docker container), then the autoscaling group terminated the EC2 instance and restarted another EC2 instance, and finally deployed the new site successfully. However, the document site was temporarily down during the redeployment and Terraform considered it as a failure and terminated the Github Action flow.

I guess the reason why these containers get killed is out-of-memory when AWS temporarily runs two Docker containers on a `t3.micro` EC2 instance. Now we have switched to `t3.small` in c7137c27e1a1d3911bb830077692bc2867198b1a, which has more memory, and should have mitigated the issue.

## Background

Currently we had configured an autoscaling group running Docker containers managed by AWS Elastic Beanstalk. However, the autoscaling group always contains a single EC2 instance because it does not support scaling up based on number of running Docker containers.

## Future work

It seems ECS on Fargate would be better than Elastic Beanstalk for our use case, because Fargate is a public cluster shared with other AWS users, which should always have available memory to deploy a new container.

The cons is that Fargate is more expensive than Elastic Beanstalk for low-traffic applications according to [the comparison](https://blog.boltops.com/2018/04/22/heroku-vs-ecs-fargate-vs-ec2-on-demand-vs-ec2-spot-pricing-comparison/#comparison-list), but it might be worth if it could save our time fighting with the private autoscaling groups.

We can try to migrate from Elastic Beanstalk to ECS on Fargate if the failures appear continuously even when we are using `t3.small` instances.

Another option is to use EKS to declare memory and CPU requirement and set up an autoscaler according to [this document](https://docs.aws.amazon.com/eks/latest/userguide/autoscaling.html), but it is over-complicated given that what we need is just to allocate a new EC2 instance during deployment.

**Comment by Atry** [comment]
#1223 did not fix this issue and it was now reverted, because there are more issues preventing the site from being deployed.

1. If an Elastic Beanstalk environment had been deployed previously, changing the settings in its related Elastic Beanstalk configuration template would not change the actual settings in that environment. There might be some workaround to force the environment settings get updated like 09ad5230d270e6d66a38b66058e179392cbaaadd, but we have not fully tested the workaround because of the other issues below
1. The existing Terraform file includes unsupported settings. According to [the comment](https://github.com/hhvm/user-documentation/blob/40c287ad998af63748016731622f70ec884af82d/.deploy/terraform/main.tf#L1-L9) the unsupported settings should be silently ignored and require manually changing the settings. However, according to the [GitHub Action log](https://github.com/hhvm/user-documentation/runs/6778690111?check_suite_focus=true), unsupported settings result in failures instead of being silently ignored in latest Terraform/AWS API.

**Comment by Atry** [comment]
## Solutions

### Option 1: Fixing existing Terraform configuration for Elastic Beanstalk

#### Pros

- We don't have to rewrite the deployment configuration and scripts

#### Cons

- We cannot automatically scale up the cluster temporarily for deployment.
- Some load balance settings, which we need, are unsupported in Elastic Beanstalk
- We don't know if there are other system limitation in Elastic Beanstalk given that it is just a wrapper of ECS and supports only a subset of features of ECS

### Option 2: Avoiding Docker
#### How?

1. Baking the site into an AMI instead a Docker image
2. Directly using Auto Scaling Group and EC2 to deploy the AMI

#### Cons

- Requires rewriting all the deployment configuration scripts 
- AMI is an Amazon proprietary format, not portable like Docker

#### Pros

- Cheaper than Fargate
- No OOM problem when deploying a version

### Option 3: ECS on Fargate 

As described in the [Future Work](https://github.com/hhvm/user-documentation/issues/1222#:~:text=running%20Docker%20containers.-,Future%20work,-It%20seems%20ECS) section

#### Cons

- Requires rewriting all the deployment configuration scripts 

#### Pros

- More expensive than EC2 or Elastic Beanstalk
- No OOM problem when deploying a version

**Comment by Atry** [comment]
Currently we use `terraform -var` to specify the Docker image. I think the usage of `-var` beats the purpose of IoC, because the idea of IaC is to consider cloud infrastructure as some "definitions" in code, which should be declarative Terraform files, not imperative. 

I would hard-code the Docker image ID in Terraform files in a new branch other than `main`. When the `main` branches change, the GitHub Action flow in `main` branch should build the Docker image and push a commit, which updates the hard-coded Docker image ID, to the branch containing Terraform files, then the pushed commit will trigger a Github Action to actually deploy the site, similar to how GitHub Pages works.

**Comment by Atry** [comment]
OK. Now I successfully deployed docs sites on Fargate.
- http://docs-prod-lb-1417095847.us-west-2.elb.amazonaws.com/
- http://docs-staging-lb-1044894882.us-west-2.elb.amazonaws.com/

The deployment files of staging and prod sites are in different branches:
- https://github.com/hhvm/user-documentation/tree/prod
- https://github.com/hhvm/user-documentation/tree/staging

**Comment by Atry** [comment]
I moved Terraform files to `deploy/xxx` branches. See #1244 for more information.

**Comment by Atry** [comment]
@AndrewDiMola I just merged #1244. Future commits to `main` branch should be automatically be deployed to http://docs-prod-main-lb-1333127108.us-west-2.elb.amazonaws.com

If everything goes well, I will change the CNAME record, letting `docs.hhvm.com` be an alias of `docs-prod-main-lb-1333127108.us-west-2.elb.amazonaws.com`

**Comment by AndrewDiMola** [comment]
@Atry I noticed that the [latest PR](https://github.com/hhvm/user-documentation/pull/1235) is merged successfully here: http://docs-prod-main-lb-1333127108.us-west-2.elb.amazonaws.com/hack/silencing-errors/error-codes but has not merged into the main production site, which is here: https://docs.hhvm.com/hack/silencing-errors/error-codes.

Just wondering if there needs to be an additional swap or overwrite. Thanks!

**Comment by AndrewDiMola** [comment]
> If everything goes well, I will change the CNAME record, letting docs.hhvm.com be an alias of docs-prod-main-lb-1333127108.us-west-2.elb.amazonaws.com

Is this why the latest PR was merged into http://docs-prod-main-lb-1333127108.us-west-2.elb.amazonaws.com/ and not https://docs.hhvm.com/? Just trying to understand. Thank you!
